lsp/src/test/kotlin/org/gradle/declarative/lsp/service/DeclarativeTextDocumentServiceForBuildScriptsTest.kt

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2025 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.declarative.lsp.service
+
+import org.junit.jupiter.api.Test
+import kotlin.io.path.readLines
+import kotlin.test.assertEquals
+
+class DeclarativeTextDocumentServiceForBuildScriptsTest : AbstractDeclarativeTextDocumentServiceTest() {
+
+    override fun script() = buildFile
+
+    @Test
+    fun `code completion inside block with file properties`() {
+        openFile(script())
+
+        assertEquals(
+            "    secrets {         }",
+            script().readLines()[4]
+        )
+
+        assertCompletion(
+            script(), 4, 16, """
+                |defaultPropertiesFile = layout.projectDirectory.file(path: String) --> defaultPropertiesFile = layout.projectDirectory.file("${'$'}{1}")${'$'}0
+                |defaultPropertiesFile = layout.settingsDirectory.file(path: String) --> defaultPropertiesFile = layout.settingsDirectory.file("${'$'}{1}")${'$'}0
+                |enabled = Boolean --> enabled = ${'$'}{1|true,false|}
+            """.trimMargin()
+        )
+    }
+
+    @Test
+    fun `code completion inside block with list properties`() {
+        openFile(script())
+
+        assertEquals(
+            "        release {         }",
+            script().readLines()[7]
+        )
+
+        assertCompletion(
+            script(), 7, 21, """
+                |baselineProfile { this: BaselineProfile } --> baselineProfile {${'\n'}|${'\t'}${'$'}0${'\n'}|}
+                |defaultProguardFiles = listOf(elements: vararg Array<ProguardFile>) --> defaultProguardFiles = listOf(${'$'}1)${'$'}0
```

Comment:
Ah, so this is where our inability to provide completion options inside the statement bites us badly.

Ideally we'd suggest `defaultProguardFiles = listOf(<caret>)` and then, once inserted, we'd suggest `proguardFile(<caret>)` in a separate completion popup. But we are not there yet.

Reply:
We will definitely be able to do that at some point. I haven't changed the completion logic that much YET, but we definitely can do it.

Reply:
It has more to do with the data that the DCL parser and resolver provide than with the completion logic in the language server.

The language server could potentially handle the special case of providing the completion `proguardFile()` in `defaultProguardFiles = listOf(<caret>)`, but that is only because the statement `defaultProguardFiles = listOf(<caret>)` is valid by itself.

Unfortunately, our current resolver will not not provide any useful resolution results for incomplete statements like `defaultProguardFiles = listOf(proguardFile("foo"), <caret>)`, which we'd like to handle just as well. This fits in the batch of work that we have long had in mind around parsing and resolving incomplete code for better IDE support.